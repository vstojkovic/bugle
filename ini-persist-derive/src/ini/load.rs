use proc_macro2::{Span, TokenStream};
use quote::{quote, quote_spanned};
use syn::{DeriveInput, Ident, Result, Type};

use super::expand_ini_impl;

pub fn expand_ini_load(input: DeriveInput) -> Result<TokenStream> {
    expand_ini_impl(input, expand_field, expand_trait)
}

fn expand_field(name: &Ident, _typ: &Type, section: TokenStream, span: Span) -> TokenStream {
    quote_spanned! { span =>
        if let Some(section) = ini.section(#section) {
            self.#name.load_in(section, "")?;
        }
    }
}

fn expand_trait(struct_name: &Ident, field_expansions: Vec<TokenStream>) -> TokenStream {
    quote! {
        #[automatically_derived]
        impl ini_persist::load::IniLoad for #struct_name {
            fn load_from_ini(&mut self, ini: &ini::Ini) -> ini_persist::Result<()> {
                use ini_persist::load::LoadProperty;
                #(#field_expansions)*
                ini_persist::Result::Ok(())
            }
        }
    }
}
