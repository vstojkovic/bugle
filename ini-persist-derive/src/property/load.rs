use proc_macro2::{Span, TokenStream};
use quote::{quote, quote_spanned};
use syn::{DeriveInput, Error, Ident, Result, Type};

use super::attr::{FieldAttr, LoadFn};
use super::expand_property_impl;

pub fn expand_load_property(input: DeriveInput) -> Result<TokenStream> {
    expand_property_impl(
        input,
        expand_field,
        expand_struct_trait,
        expand_repr_variant_match,
        expand_named_variant_match,
        expand_enum_trait,
    )
}

fn expand_field(
    name: &Ident,
    _typ: &Type,
    key: TokenStream,
    attr: FieldAttr,
    span: Span,
) -> TokenStream {
    match attr.load_fn {
        None => quote_spanned! { span =>
            self.#name.load_in(section, #key)?;
        },
        Some(LoadFn::InPlace(path)) => quote_spanned! { span =>
            #path(&mut self.#name, section, #key)?;
        },
        Some(LoadFn::Constructed(path)) => quote_spanned! { span =>
            if let Some(value) = #path(section, #key)? {
                self.#name = value;
            }
        },
        Some(LoadFn::Parsed(path)) => quote_spanned! { span =>
            if let Some(value) = section.get(#key) {
                self.#name = #path(value)?;
            }
        },
    }
}

fn expand_struct_trait(
    struct_name: &Ident,
    field_expansions: Vec<Result<TokenStream>>,
) -> TokenStream {
    let load_calls = field_expansions
        .into_iter()
        .map(|expansion| expansion.unwrap_or_else(Error::into_compile_error));
    quote! {
        #[automatically_derived]
        impl ini_persist::load::LoadProperty for #struct_name {
            fn load_in(&mut self, section: &ini::Properties, key: &str) -> ini_persist::Result<()> {
                #(#load_calls)*
                ini_persist::Result::Ok(())
            }
        }
    }
}

fn expand_named_variant_match(name: &Ident, enum_name: &Ident, span: Span) -> TokenStream {
    quote_spanned!(span => stringify!(#name) => #enum_name::#name)
}

fn expand_repr_variant_match(name: &Ident, enum_name: &Ident, span: Span) -> TokenStream {
    quote_spanned!(span => Discriminants::#name => #enum_name::#name)
}

fn expand_enum_trait(
    enum_name: &Ident,
    repr_type: Option<&Ident>,
    prelude: TokenStream,
    match_arms: Vec<TokenStream>,
) -> TokenStream {
    let parse_repr = repr_type
        .map(|repr_type| {
            quote! {
                let value = #repr_type::parse(value)?;
            }
        })
        .unwrap_or_default();
    quote! {
        #[automatically_derived]
        impl ini_persist::load::ParseProperty for #enum_name {
            fn parse(value: &str) -> ini_persist::Result<Self> {
                #prelude
                #parse_repr
                ini_persist::Result::Ok(match value {
                    #(#match_arms,)*
                    _ => return ini_persist::Result::Err(ini_persist::Error::invalid_value(
                        format!("invalid value: {}", value)
                    ))
                })
            }
        }
    }
}
