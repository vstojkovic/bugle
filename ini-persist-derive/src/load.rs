use proc_macro2::{Span, TokenStream};
use quote::{quote, quote_spanned};
use syn::spanned::Spanned;
use syn::{Data, DataStruct, DeriveInput, Error, Fields, Ident, Result};

use crate::attr::{FieldAttr, LoadFn, StructAttr};

pub fn expand_load(input: DeriveInput) -> Result<TokenStream> {
    let fields = match &input.data {
        Data::Struct(DataStruct {
            fields: Fields::Named(fields),
            ..
        }) => &fields.named,
        _ => panic!("IniLoad can only be derived on a struct with named fields"),
    };

    let attr = StructAttr::from_ast(input.attrs.iter())?;
    let struct_name = input.ident;
    let section_name = attr
        .section
        .unwrap_or_else(|| Some(struct_name.to_string()));
    let section_quote = match section_name {
        Some(name) => quote!(Some(#name)),
        None => quote!(None::<String>),
    };

    let load_calls = fields
        .into_iter()
        .map(|field| {
            let attr = FieldAttr::from_ast(field.attrs.iter())?;
            let field_name = field.ident.as_ref().unwrap();
            let span = field.ty.span();

            if attr.flatten.is_some() {
                expand_flattened_field(field_name, span)
            } else {
                expand_field(field_name, attr, span)
            }
        })
        .map(|result| result.unwrap_or_else(Error::into_compile_error));

    let output = quote! {
        #[automatically_derived]
        impl ini_persist::load::IniLoad for #struct_name {
            fn load_from_ini(&mut self, ini: &ini::Ini) -> ini_persist::Result<()> {
                use ini_persist::load::Property;
                if let Some(props) = ini.section(#section_quote) {
                    #(#load_calls)*
                }
                Ok(())
            }
        }
    };

    Ok(output)
}

fn expand_field(name: &Ident, attr: FieldAttr, span: Span) -> Result<TokenStream> {
    let key = attr.key.unwrap_or_else(|| name.to_string());
    Ok(match attr.load_fn {
        None => quote_spanned! { span =>
            self.#name.load_in(props, #key)?;
        },
        Some(LoadFn::InPlace(path)) => quote_spanned! { span =>
            #path(&mut self.#name, props, #key)?;
        },
        Some(LoadFn::Constructed(path)) => quote_spanned! { span =>
            if let Some(value) = #path(props, #key)? {
                self.#name = value;
            }
        },
        Some(LoadFn::Parsed(path)) => quote_spanned! { span =>
            if let Some(value) = props.get(#key) {
                self.#name = #path(value)?;
            }
        },
    })
}

fn expand_flattened_field(name: &Ident, span: Span) -> Result<TokenStream> {
    Ok(quote_spanned! { span =>
        self.#name.load_from_ini(ini)?;
    })
}
