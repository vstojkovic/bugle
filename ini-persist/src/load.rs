use ini::{Ini, Properties};

use crate::Result;

pub trait IniLoad {
    fn load_from_ini(&mut self, ini: &Ini) -> Result<()>;
}

pub trait Property {
    fn load_in(&mut self, section: &Properties, key: &str) -> Result<()>;
}

pub trait ConstructedProperty: Sized {
    fn load(section: &Properties, key: &str) -> Result<Option<Self>>;
}

impl<P: ConstructedProperty> Property for P {
    fn load_in(&mut self, section: &Properties, key: &str) -> Result<()> {
        if let Some(value) = P::load(section, key)? {
            *self = value;
        }
        Ok(())
    }
}

pub trait ParsedProperty: Sized {
    fn parse(text: &str) -> Result<Self>;
}

impl<P: ParsedProperty> ConstructedProperty for P {
    fn load(section: &Properties, key: &str) -> Result<Option<Self>> {
        match section.get(key) {
            Some(text) => Ok(Some(P::parse(text)?)),
            None => Ok(None),
        }
    }
}

impl ParsedProperty for String {
    fn parse(text: &str) -> Result<Self> {
        Ok(text.to_string())
    }
}

macro_rules! impl_from_str_properties {
    ($($type:ty),+ $(,)?) => {
        $(
        impl ParsedProperty for $type {
            fn parse(text: &str) -> $crate::Result<Self> {
                use std::str::FromStr;
                Self::from_str(text).map_err(|err| {
                    $crate::error::Error::invalid_type(format!(
                        concat!("failed to parse ", stringify!($type), " from: {}"),
                        text
                    ))
                    .with_cause(err)
                })
            }
        }
        )+
    };
}

impl_from_str_properties!(
    i8, i16, i32, i64, i128, isize, u8, u16, u32, u64, u128, usize, f32, f64, bool, char,
);
