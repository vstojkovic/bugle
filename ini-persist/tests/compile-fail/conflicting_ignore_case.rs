use ini_persist::load::LoadProperty;

#[derive(LoadProperty)]
#[ini(repr)]
#[ini(ignore_case)]
enum Foo {
    Arglebargle,
    GlopGlyf,
}

#[derive(LoadProperty)]
#[ini(ignore_case)]
#[ini(repr)]
enum Bar {
    OlleBolle,
    SnopSnyf,
}

fn main() {
}
