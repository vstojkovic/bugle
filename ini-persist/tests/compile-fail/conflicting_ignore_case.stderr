error: cannot specify both `repr` and `ignore_case`
 --> tests/compile-fail/conflicting_ignore_case.rs:5:7
  |
5 | #[ini(ignore_case)]
  |       ^^^^^^^^^^^

error: cannot specify both `repr` and `ignore_case`
  --> tests/compile-fail/conflicting_ignore_case.rs:13:7
   |
13 | #[ini(repr)]
   |       ^^^^
